zero/middleware/auth.py

```diff
@@ -0,0 +1,41 @@
+"""Example middleware for decoding encrypted JWTs on each request."""
+
+from typing import Callable, Iterable, Tuple
+import jwt
+from zero import config
+from arxiv.base.middleware import BaseMiddleware
+
+
+class ExampleAuthMiddleware(BaseMiddleware):
+    """
+    An example of middleware to handle auth information on requests.
+
+    Before the request is handled by the application, the ``Authorization``
+    header is parsed for an encrypted JWT. If successfully decrypted,
+    information about the user and their authorization scope is attached
+    to the request.
+
+    This can be accessed in the application via
+    ``flask.request.environ['auth']``.  If Authorization header was not
+    included, or if the JWT could not be decrypted, then that value will be
+    ``None``.
+    """
+
+    def before(self, environ: dict, start_response: Callable) \
+            -> Tuple[dict, Callable]:
+        """Parse the ``Authorization`` header in the response."""
+        auth_header = environ.get('HTTP_AUTHORIZATION')
+        environ['auth'] = None
+        if not auth_header:
+            return environ, start_response
+        try:
+            decoded = jwt.decode(auth_header, config.JWT_SECRET,
+                                 algorithms=['HS256'])
+        except jwt.exceptions.DecodeError:  # type: ignore
+            return environ, start_response
```

Comment:
If I understand this correctly, the middleware does **NOT** handle a failed JWT token decode with a 401 or 403 or anything like that -that's the role of the application logic to deal with `environ['auth']['scope']` and `environ['auth']['user']` correctly?

Reply:
Right. The idea is that each service should focus on defining and enforcing authorization rules, and leave the authentication bits outside the application.

This is just an example; probably this will be implemented more comprehensively in the arxiv-accounts project as an installable package, so that it is consistent across all services that expose sensitive endpoints.